sdks/go/pkg/beam/core/runtime/graphx/translate.go

```diff
@@ -0,0 +1,651 @@
+// limitations under the License.
+
+package graphx
+
+import (
+	"fmt"
+
+	"strconv"
+	"strings"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/window"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx/v1"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/protox"
+	fnpb "github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+	pb "github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1"
+	"github.com/golang/protobuf/proto"
+)
+
+const (
+	// Model constants
+
+	urnImpulse = "urn:beam:transform:impulse:v1" // to appear?
+	urnParDo   = "urn:beam:transform:pardo:v1"
+	urnFlatten = "urn:beam:transform:flatten:v1"
+	urnGBK     = "urn:beam:transform:groupbykey:v1"
+	urnCombine = "urn:beam:transform:combine:v1"
+	urnWindow  = "urn:beam:transform:window:v1"
+
+	urnDataSource = "urn:org.apache.beam:source:runner:0.1"
+	urnDataSink   = "urn:org.apache.beam:sink:runner:0.1"
+
+	urnGlobalWindowsWindowFn = "beam:windowfn:global_windows:v0.1"
+
+	// SDK constants
+
+	urnDoFn = "urn:beam:dofn:javasdk:0.1" // TODO: use "urn:beam:go:transform:dofn:v1"
+)
+
+// TODO(herohde) 11/6/2017: move some of the configuration into the graph during construction.
+
+// Options
+type Options struct {
+	// ContainerImageURL is the default environment container image.
+	ContainerImageURL string
+}
+
+// Marshal converts a graph to a model pipeline.
+func Marshal(edges []*graph.MultiEdge, opt *Options) (*pb.Pipeline, error) {
+	tree := NewScopeTree(edges)
+	EnsureUniqueNames(tree)
+
+	m := newMarshaller(opt.ContainerImageURL)
+
+	var roots []string
+	for _, edge := range tree.Edges {
+		roots = append(roots, m.addMultiEdge(edge))
+	}
+	for _, t := range tree.Children {
+		roots = append(roots, m.addScopeTree(t))
+	}
+
+	p := &pb.Pipeline{
+		Components:       m.build(),
+		RootTransformIds: roots,
+	}
+	return p, nil
+}
+
+type marshaller struct {
+	imageURL string
+
+	transforms   map[string]*pb.PTransform
+	pcollections map[string]*pb.PCollection
+	windowing    map[string]*pb.WindowingStrategy
+	environments map[string]*pb.Environment
+
+	coders *CoderMarshaller
+}
+
+func newMarshaller(imageURL string) *marshaller {
+	return &marshaller{
+		imageURL:     imageURL,
+		transforms:   make(map[string]*pb.PTransform),
+		pcollections: make(map[string]*pb.PCollection),
+		windowing:    make(map[string]*pb.WindowingStrategy),
+		environments: make(map[string]*pb.Environment),
+		coders:       NewCoderMarshaller(),
+	}
+}
+
+func (m *marshaller) build() *pb.Components {
+	return &pb.Components{
+		Transforms:          m.transforms,
+		Pcollections:        m.pcollections,
+		WindowingStrategies: m.windowing,
+		Environments:        m.environments,
+		Coders:              m.coders.Build(),
+	}
+}
+
+func (m *marshaller) addScopeTree(s *ScopeTree) string {
+	id := scopeID(s.Scope.Scope)
+	if _, exists := m.transforms[id]; exists {
+		return id
+	}
+
+	var sub []string
```

Comment:
<!--thread_id:cc_155329520_t; commit:c5876af98386a806693a64b6dc5905f2397c35fb; resolved:1-->
<!--section:context-quote-->
> **lostluck** wrote:
> sub is a list of edges? I don't know what the sub list is for from the context though, substituting? subtraction? sublimation?

<!--section:body-->
Changed to subtransforms.